sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala

```diff
@@ -246,6 +267,197 @@ object RewritePredicateSubquery extends Rule[LogicalPlan] with PredicateHelper {
+
+      // Get all aggregate expressions associated with interesting expressions.
+      val aggregateExprs = inSubqueryMapping.flatMap(_._2)
+      // Create aliases for each above aggregate expression. We can't use the aggregate
+      // expressions directly in the new Aggregate node because Aggregate.aggregateExpressions
+      // has the type Seq[NamedExpression].
+      val aggregateExprAliases = aggregateExprs.map(a => Alias(a, toPrettySQL(a))())
+      // Create a mapping from each aggregate expression to its alias.
+      val aggregateExprAliasMap = aggregateExprs.zip(aggregateExprAliases).toMap
+      // Create attributes from those aliases of aggregate expressions. These attributes
+      // will be used in the new Project node to refer to the aliased aggregate expressions
+      // in the new Aggregate node.
+      val aggregateExprAttrs = aggregateExprAliases.map(_.toAttribute)
+      // Create a mapping from aggregate expressions to attributes. This will be
+      // used when patching the interesting expressions after they are pulled up
+      // into the new Project node: aggregate expressions will be replaced by attributes.
+      val aggregateExprAttrMap = aggregateExprs.zip(aggregateExprAttrs).toMap
+
+      // Create an Aggregate node without the interesting expressions, just
+      // the associated aggregate expressions plus any other group-by or aggregate expressions
+      // that were not involved in the interesting expressions.
+      val newAggregateExpressions = a.aggregateExpressions.flatMap {
+        // If this expression contains IN-subqueries with aggregates in the left-hand
+        // operand, replace with just the aggregates.
+        case ae: Expression if inSubqueryMap.contains(ae) =>
+          // Replace the expression with an aliased aggregate expression.
+          inSubqueryMap(ae).map(aggregateExprAliasMap(_))
+        case ae => Seq(ae)
+      }
+      val newAggregate = a.copy(aggregateExpressions = newAggregateExpressions)
+
+      // Create a projection with the IN-subquery expressions that contain aggregates, replacing
+      // the aggregate expressions with attribute references to the output of the new Aggregate
+      // operator. Also include the other output of the Aggregate operator.
+      val projList = a.aggregateExpressions.map {
+        // If this expression contains an IN-subquery that uses an aggregate, we
+        // need to do something special
+        case ae: Expression if inSubqueryMap.contains(ae) =>
+          ae.transform {
+            // Patch any aggregate expression with its corresponding attribute.
+            case a: AggregateExpression => aggregateExprAttrMap(a)
+          }.asInstanceOf[NamedExpression]
+        case ae => ae.toAttribute
+      }
+      val newProj = Project(projList, newAggregate)
+
+      // Reapply this rule, but now with all interesting expressions
+      // from Aggregate.aggregateExpressions pulled up into a Project node.
+      apply(newProj)
```

Comment:
This reminds me of the rule `RewriteWithExpression`, which also needs to rewrite `Aggregate` first. We should not call `apply` here in the middle of plan traveral, as `apply` transforms the plan again, and leads to O(n^2) complexity. Instead of, we should also add a util function that rewrites `UnaryNode` (not transforms the full tree) and call it here and the original case match for `UnaryNode`.